File: AlgorithmStudy/11주차/BOJ2644/김재환2644.py
import sys
from collections import deque

# 빠른 입력
input = sys.stdin.readline

# 1. 전체 사람 수
n = int(input().strip())

# 2. 촌수 계산할 두 사람
a, b = map(int, input().split())

# 3. 관계 개수
m = int(input().strip())

# 그래프 인접 리스트
graph = [[] for _ in range(n + 1)]

for _ in range(m):
    x, y = map(int, input().split())
    # 부모/자식 관계는 양방향
    graph[x].append(y)
    graph[y].append(x)

# 방문 체크 및 촌수 계산 배열
visited = [0] * (n + 1)

# BFS 큐
queue = deque()
queue.append(a)
visited[a] = 1  # 시작 노드 촌수 base = 1

result = -1

while queue:
    current = queue.popleft()

    # 목표에 도달하면 촌수 계산 (실제 촌수 = visited - 1)
    if current == b:
        result = visited[current] - 1
        break

    # 연결된 사람들 탐색
    for nxt in graph[current]:
        if visited[nxt] == 0:
            visited[nxt] = visited[current] + 1
            queue.append(nxt)

# 연결되지 않았으면 -1, 아니면 계산된 촌수 출력
print(result)

File: AlgorithmStudy/11주차/BOJ2644/김재환2644.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws Exception {
        // 빠른 입력 처리
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        // 1. 전체 사람 수
        int n = Integer.parseInt(br.readLine().trim());

        // 2. 촌수를 구할 두 사람 번호 입력
        st = new StringTokenizer(br.readLine());
        int personA = Integer.parseInt(st.nextToken()); // 시작점
        int personB = Integer.parseInt(st.nextToken()); // 목표점

        // 3. 부모 자식 관계 수
        int m = Integer.parseInt(br.readLine().trim());

        /*
         * 가족 관계를 저장할 인접 리스트 그래프
         * 각 사람 번호 최대 n까지 존재하므로
         * index 1부터 사용하기 위해 크기를 n+1로 설정
         */
        ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            graph.add(new ArrayList<>());
        }

        // 부모-자식 관계 입력
        // 이 문제는 부모와 자식이 주어지지만,
        // 실제로는 양방향 관계(부모↔자식)이므로 둘 다 추가
        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int p1 = Integer.parseInt(st.nextToken());
            int p2 = Integer.parseInt(st.nextToken());

            // 양방향 연결: p1 ↔ p2
            graph.get(p1).add(p2);
            graph.get(p2).add(p1);
        }

        // 방문 여부 및 촌수 저장 배열
        // visited[i] = i번 사람까지의 촌수
        int[] visited = new int[n + 1];

        // BFS 수행
        Queue<Integer> queue = new LinkedList<>();
        queue.add(personA);       // 시작 노드 push
        visited[personA] = 1;      // 촌수 계산을 위해 1로 시작

        // 촌수 결과
        int result = -1;

        while (!queue.isEmpty()) {
            int current = queue.poll();

            // 목표 인물에 도달하면 종료
            if (current == personB) {
                // 실제 촌수는 visited 값 - 1
                result = visited[current] - 1;
                break;
            }

            // 인접한 모든 관계 노드 탐색
            for (int next : graph.get(current)) {
                // 아직 방문하지 않았다면
                if (visited[next] == 0) {
                    // 현재 촌수 + 1로 저장
                    visited[next] = visited[current] + 1;
                    queue.add(next);
                }
            }
        }

        // 출력 (연결이 없다면 result = -1 그대로)
        System.out.println(result);
    }
}
